tokio/src/util/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,783 @@
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_sorted(&mut self, node: &mut ListNode<T>)
+    where
+        T: PartialOrd,
+    {
+        if self.head.is_none() {
+            // First node in the list
+            self.head = Some(node.into());
+            self.tail = Some(node.into());
+            return;
+        }
+
+        let mut prev: Option<NonNull<ListNode<T>>> = None;
+        let mut current = self.head;
+
+        while let Some(mut current_node) = current {
+            if node.data < current_node.as_ref().data {
+                // Need to insert before the current node
+                current_node.as_mut().prev = Some(node.into());
+                match prev {
+                    Some(mut prev) => {
+                        prev.as_mut().next = Some(node.into());
+                    }
+                    None => {
+                        // We are inserting at the beginning of the list
+                        self.head = Some(node.into());
+                    }
+                }
+                node.next = current;
+                node.prev = prev;
+                return;
+            }
+            prev = current;
+            current = current_node.as_ref().next;
+        }
+
+        // We looped through the whole list and the nodes data is bigger or equal
+        // than everything we found up to now.
+        // Insert at the end. Since we checked before that the list isn't empty,
+        // tail always has a value.
+        node.prev = self.tail;
+        node.next = None;
+        self.tail.as_mut().unwrap().as_mut().next = Some(node.into());
+        self.tail = Some(node.into());
+    }
+
+    /// Returns the first node in the linked list without removing it from the list
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    /// The returned pointer is only guaranteed to be valid as long as the list
+    /// is not mutated
+    pub fn peek_first(&self) -> Option<&mut ListNode<T>> {
+        // Safety: When the node was inserted it was promised that it is alive
+        // until it gets removed from the list.
+        // The returned node has a pointer which constrains it to the lifetime
+        // of the list. This is ok, since the Node is supposed to outlive
+        // its insertion in the list.
+        unsafe {
+            self.head
+                .map(|mut node| &mut *(node.as_mut() as *mut ListNode<T>))
+        }
+    }
+
+    /// Returns the last node in the linked list without removing it from the list
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    /// The returned pointer is only guaranteed to be valid as long as the list
+    /// is not mutated
+    pub fn peek_last(&self) -> Option<&mut ListNode<T>> {
```

Comment:
IMO these methods should probably return a pinned node, if possible. This would prevent callers from using `mem::replace` or `mem::swap` to break the safety invariant that the nodes addresses remain stable in safe code. 

If these were pinned, then callers could only move them using `Pin::get_unchecked_mut`, which is unsafe.